Mathlib/Topology/MetricSpace/UniformConvergence.lean

```diff
@@ -0,0 +1,280 @@
+      exact ⟨δ, hδ, by simpa [iSup_lt_iff] using fun f g h ↦ ⟨δ, hδε, h⟩⟩
+    · simp only [gt_iff_lt, iSup_le_iff, Filter.le_principal_iff, Filter.mem_principal,
+        Set.setOf_subset_setOf, Prod.forall]
+      exact fun ε hε f g h x hx ↦ (le_iSup₂ (f := fun x _ ↦ edist (f x) (g x)) x hx).trans h.le
+
+lemma edist_def (f g : α →ᵤ[𝔖] β) :
+    edist f g = ⨆ x ∈ ⋃₀ 𝔖, edist (toFun 𝔖 f x) (toFun 𝔖 g x) :=
+  rfl
+
+lemma edist_def' (f g : α →ᵤ[𝔖] β) :
+    edist f g = ⨆ s ∈ 𝔖, ⨆ x ∈ s, edist (toFun 𝔖 f x) (toFun 𝔖 g x) := by
+  simp [edist_def, iSup_and, iSup_comm (ι := α)]
+
+lemma lipschitzWith_iff {f : γ → α →ᵤ[𝔖] β} {K : ℝ≥0} :
+    LipschitzWith K f ↔ ∀ c ∈ ⋃₀ 𝔖, LipschitzWith K (fun x ↦ toFun 𝔖 (f x) c) := by
+  simp [LipschitzWith, edist_def, toFun, ofFun]
+  tauto
+
+lemma lipschitzOnWith_iff {f : γ → α →ᵤ[𝔖] β} {K : ℝ≥0} {s : Set γ} :
+    LipschitzOnWith K f s ↔ ∀ c ∈ ⋃₀ 𝔖, LipschitzOnWith K (fun x ↦ toFun 𝔖 (f x) c) s := by
+  simp [lipschitzOnWith_iff_restrict, lipschitzWith_iff]
+  rfl
+
+lemma lipschitzWith_eval (x : α) (hx : x ∈ ⋃₀ 𝔖) :
+    LipschitzWith 1 (fun f : α →ᵤ[𝔖] β ↦ toFun 𝔖 f x) := by
+  intro f g
+  simpa only [ENNReal.coe_one, one_mul] using
+    le_iSup₂ (f := fun y _ ↦ edist (toFun 𝔖 f y) (toFun 𝔖 g y)) x hx
+
+lemma lipschitzWith_one_ofFun_toFun :
+    LipschitzWith 1 (ofFun 𝔖 ∘ UniformFun.toFun : (α →ᵤ β) → (α →ᵤ[𝔖] β)) :=
+  lipschitzWith_iff.mpr fun _ _ ↦ UniformFun.lipschitzWith_eval _
+
+lemma lipschitzWith_one_ofFun_toFun' [Finite 𝔗] (h : ⋃₀ 𝔖 ⊆ ⋃₀ 𝔗) :
+    LipschitzWith 1 (ofFun 𝔖 ∘ toFun 𝔗 : (α →ᵤ[𝔗] β) → (α →ᵤ[𝔖] β)) :=
+  lipschitzWith_iff.mpr fun x hx ↦ lipschitzWith_eval x (h hx)
+
+lemma lipschitzWith_restrict (s : Set α) (hs : s ∈ 𝔖)  :
+    LipschitzWith 1 (UniformFun.ofFun ∘ s.restrict ∘ toFun 𝔖 : (α →ᵤ[𝔖] β) → (s →ᵤ β)) :=
+  UniformFun.lipschitzWith_iff.mpr fun x ↦ lipschitzWith_eval _ ⟨s, hs, x.2⟩
+
+lemma isometry_restrict (s : Set α) :
+    Isometry (UniformFun.ofFun ∘ s.restrict ∘ toFun {s} : (α →ᵤ[{s}] β) → (s →ᵤ β)) := by
+  simp [Isometry, edist_def, UniformFun.edist_def, iSup_subtype]
+
+end EMetric
+
+section Metric
+
+variable [Finite 𝔖] [PseudoMetricSpace β]
+
+noncomputable instance [BoundedSpace β] : PseudoMetricSpace (α →ᵤ[𝔖] β) :=
+  PseudoEMetricSpace.toPseudoMetricSpaceOfDist
+    (fun f g ↦ ⨆ x ∈ ⋃₀ 𝔖, dist (toFun 𝔖 f x) (toFun 𝔖 g x))
+    (fun _ _ ↦ by
+      have := BoundedSpace.bounded_univ (α := β) |>.ediam_ne_top.lt_top
+      refine (iSup₂_le fun x _ ↦ EMetric.edist_le_diam_of_mem ?_ ?_).trans_lt this |>.ne
+      all_goals trivial)
+    (fun _ _ ↦ by
+      simp only [dist_edist, edist_def, iSup_sigma']
+      rw [ENNReal.toReal_iSup]
+      · congr!
+        rw [ENNReal.toReal_iSup]
+        exact (fun _ ↦ edist_ne_top _ _)
+      · have := BoundedSpace.bounded_univ (α := β) |>.ediam_ne_top.lt_top
+        refine fun x ↦ lt_of_le_of_lt (iSup_le fun hx ↦ ?_) this |>.ne
+        exact EMetric.edist_le_diam_of_mem (by trivial) (by trivial))
```

Comment:
Here you're making your life extra hard by using a conditional supremum with real values... Here's a golf:
```suggestion
      simp only [dist_edist, edist_def, ← ENNReal.toReal_iSup (fun _ ↦ edist_ne_top _ _)]
      rw [ENNReal.toReal_iSup]
      have := BoundedSpace.bounded_univ (α := β) |>.ediam_ne_top.lt_top
      refine fun x ↦ lt_of_le_of_lt (iSup_le fun hx ↦ ?_) this |>.ne
      exact EMetric.edist_le_diam_of_mem (by trivial) (by trivial))
```

That said, I wonder if the distances (and maybe even the edistances) should first be defined as they are for BoundedContinuousFunction, i.e as the best positive bound. This has the advantage of always existing, so we don't really on the explicit junk values. What do you think?

Reply:
:thinking: maybe for `dist`, but not `edist` as in the latter case all the lemmas you want come for free from `iSup` on complete lattices. But I'm a bit hesitant to bother changing it for `dist` since this should rarely apply (i.e., it only appears if `β` is a bounded space).

Reply:
My point was that it's better for dist, and for edist it doesn't change anything, so if we make the change I would favor consistency. But let's stick with this.

Reply:
Yeah, somehow I think that, although when `β` is bounded we technically have a `dist`, it's still cursed (whether we use the sup or the inf definition) and you should always use `edist`.